packages/ui-components/src/components/toolbar.tsx

```diff
@@ -503,6 +503,11 @@ export namespace ToolbarButtonComponent {
     tooltip?: string;
     onClick?: () => void;
     enabled?: boolean;
+    ariaLabel?: string;
+    pressed?: boolean;
+    pressedIcon?: LabIcon.IMaybeResolvable;
+    pressedTooltip?: string;
+    disabledTooltip?: string;
```

Comment:
Hey I though a bit more about the API for this. And I think it will be better to not complicate more the `ToolbarButtonComponent`. I would suggest enhancing the `CommandToolbarButton` and use the command attribute like `isToggled` to get the pressed status and `isToggleable` (see [doc](https://github.com/jupyterlab/lumino/blob/97e9393124c4effc66bedc55aa6bd9198df94309/packages/commands/src/index.ts#L331)). Then we can let the developer responsible to update the label and the icon depending on the state of the command.

The `CommandToolbarButton` will need to update the underlying `ToolbarButtonComponent` if the command is toggleable and the user has clicked on the button.

Reply:
ToolbarButtonComponent would have to still support all these properties. Correct? I am not familiar with the command interface, so you have to fill me in a little bit more. Is there an existing example that I can follow, or better if we can discuss offline.

Reply:
Hey @3coins 

Here is the current code for the `CommandToolbarButtonComponent` function:

```js
  return (
    <UseSignal
      signal={props.commands.commandChanged}
      shouldUpdate={(sender, args) =>
        (args.id === props.id && args.type === 'changed') ||
        args.type === 'many-changed'
      }
    >
      {() => <ToolbarButtonComponent {...Private.propsFromCommand(props)} />}
    </UseSignal>
  );
```

What I'm proposing is to have something like that:

```js
  const [buttonProps, setButtonProps] = React.useState(Private.propsFromCommand(props));
  
  props.commands.commandExecuted(command => {
    if(command.id === props.id){
      setButtonProps(Private.propsFromCommand(props));
    }
  });
  
  return (
    <UseSignal
      signal={props.commands.commandChanged}
      shouldUpdate={(sender, args) =>
        (args.id === props.id && args.type === 'changed') ||
        args.type === 'many-changed'
      }
    >
      {() => <ToolbarButtonComponent {...buttonProps} />}
    </UseSignal>
  );
```

The `propsFromCommand` will need to return indeed a new property for the `pressed` status of the button. But this will be the result of introspecting `commands.isToggled(command_id)`.

Reply:
It will be the responsibility of the `icon` command attribute to change with the `pressed` status (same for the `label` and the `caption`).

Reply:
Note: It looks like a similar approach was being used in #9389.

Reply:
@fcollonval 
Thanks for code snippets.

I found the command interface not intuitive at all to use, and feel that most developers will find it much harder to implement, or spend a lot of time figuring out the right structure to use it. On the contrary, setting the properties directly to the widget was much simpler and direct to follow. 

I have been trying to implement this for the last couple of days and ran into a few snags, let me know if you have any resolution. Here is what I have so far in addition to what you had suggested, this is inside the `handler.ts` for debugger.

```javascript
const debugButton = this._iconButtons[widget.id];
let enabled = false;
if(!debugButton) {
  this._commands = new CommandRegistry();
  this._commands.addCommand(COMMAND_ID, {
    execute: async args => {
      await toggleDebugging(); // this is an async call, but execute in CommandRegistry calls the function directly than using await
    },
    icon: args => {
      return this._commands.isToggled(COMMAND_ID) ? bugDotIcon : bugIcon;
    },
    caption: args => {
      if(!this._commands.isEnabled(COMMAND_ID)) {
        return "Select a kernel with debug support to enable debugging";
      }
      return this._commands.isToggled(COMMAND_ID) ? "Disable Debugging" : "Enable Debugging";
    },
    isToggled: () => {
      return isDebuggerOn();
    },
    isEnabled: async () => {
      const enabled = await this._service.isAvailable(connection) // same problem as execute
      return enabled;
    },
    isToggleable: true
  })
  this._iconButtons[widget.id] = new CommandToolbarButton({
    commands: this._commands,
    id: COMMAND_ID
  });
  widget.toolbar.insertBefore('kernelName', 'debugger-icon', this._iconButtons[widget.id]! as Widget);
} else {
  this._commands.notifyCommandChanged(COMMAND_ID); // Is this the the right API to update the command?
}
```
Here are my queries:
1. Toggle function in debugger is async, but the execute in CommandRegistry calls the function directly (doesn't use await).  The side effect of this is that the visual updates to the button are out of sync by one cycle.
2. Same issue as above for checking if debugger is enabled.
3. One way to resolve the above is keeping these states in the handler and post an update to Command by directly emitting a change. I am not sure what API to use here, is notifyCommandChanged appropriate here?.

Reply:
Hey thanks for trying out. So let keep the current implementation that fit the purpose and have a correct rendering lifecycle.